adm/src/database/lmdb.rs

```diff
@@ -0,0 +1,161 @@
+
+pub struct LmdbContext {
+    pub env: lmdb::Environment,
+}
+
+impl LmdbContext {
+    pub fn new(filepath: &Path, indexes: u32, size: Option<usize>) -> Result<Self, DatabaseError> {
+        let flags = lmdb::open::MAPASYNC | lmdb::open::WRITEMAP | lmdb::open::NOSUBDIR;
+
+        let filepath_str = filepath.to_str()
+            .ok_or(DatabaseError::InitError(format!("Invalid filepath: {:?}", filepath)))?;
+
+        let mut builder = lmdb::EnvBuilder::new().map_err(|err|
+            DatabaseError::InitError(format!("Failed to initialize environment: {}", err)))?;
+        builder.set_maxdbs(indexes + 1).map_err(|err|
+            DatabaseError::InitError(format!("Failed to set MAX_DBS: {}", err)))?;
+        builder.set_mapsize(size.unwrap_or(DEFAULT_SIZE)).map_err(|err|
+            DatabaseError::InitError(format!("Failed to set MAP_SIZE: {}", err)))?;
+
+        let env = unsafe {
+                builder.open(filepath_str, flags, 0o600)
+                    .map_err(|err|
+                        DatabaseError::InitError(format!("Database not found: {}", err)))
+        }?;
+        Ok(LmdbContext{
+            env: env
+        })
+    }
+}
+
+pub struct LmdbDatabase<'e> {
+    ctx:     &'e LmdbContext,
+    main:    lmdb::Database<'e>,
+    indexes: HashMap<String, lmdb::Database<'e>>,
+}
+
+impl<'e> LmdbDatabase<'e> {
+    pub fn new(ctx: &'e LmdbContext, indexes: &[&str]) -> Result<Self, DatabaseError> {
+        let main = lmdb::Database::open(
+            &ctx.env, Some("main"), &lmdb::DatabaseOptions::new(lmdb::db::CREATE)
+        ).map_err(|err|
+            DatabaseError::InitError(format!("Failed to open database: {:?}", err)))?;
+
+        let mut index_dbs = HashMap::with_capacity(indexes.len());
+        for name in indexes {
+            let db = lmdb::Database::open(
+                &ctx.env, Some(name), &lmdb::DatabaseOptions::new(lmdb::db::CREATE)
+            ).map_err(|err|
+                DatabaseError::InitError(
+                    format!("Failed to open database: {:?}", err)))?;
+            index_dbs.insert(String::from(*name), db);
+        }
+        Ok(LmdbDatabase{
+            ctx: ctx,
+            main: main,
+            indexes: index_dbs,
+        })
+    }
+
+    pub fn reader(&self, index: Option<&str>) -> Result<LmdbDatabaseReader, DatabaseError> {
+        let txn = lmdb::ReadTransaction::new(&self.ctx.env).map_err(|err|
+            DatabaseError::ReaderError(format!("Failed to create reader: {}", err)))?;
+        let db = match index {
+            None => Ok(&self.main),
+            Some(index) => self.indexes.get(index).ok_or(
+                DatabaseError::ReaderError(format!("Not an index: {}", index))),
+        }?;
+        Ok(LmdbDatabaseReader{
+            db: db,
+            txn: txn,
+        })
+    }
+
+    pub fn writer(&self, index: Option<&str>) -> Result<LmdbDatabaseWriter, DatabaseError> {
+        let txn = lmdb::WriteTransaction::new(&self.ctx.env).map_err(|err|
```

Comment:
Need a way to have a single transaction for multiply writers (for the case of updating multiple indexes per record in a single transaction).